core/src/operations/mod.rs

```diff
@@ -1,131 +1,242 @@
+    pub const OPCODE_DUP1: u8 = 0b0011_0010;
+    pub const OPCODE_DUP2: u8 = 0b0011_0011;
+    pub const OPCODE_DUP3: u8 = 0b0011_0100;
+    pub const OPCODE_DUP4: u8 = 0b0011_0101;
+    pub const OPCODE_DUP5: u8 = 0b0011_0110;
+    pub const OPCODE_DUP6: u8 = 0b0011_0111;
+    pub const OPCODE_DUP7: u8 = 0b0011_1000;
+    pub const OPCODE_DUP9: u8 = 0b0011_1001;
+    pub const OPCODE_DUP11: u8 = 0b0011_1010;
+    pub const OPCODE_DUP13: u8 = 0b0011_1011;
+    pub const OPCODE_DUP15: u8 = 0b0011_1100;
+    pub const OPCODE_ADVPOP: u8 = 0b0011_1101;
+    pub const OPCODE_SDEPTH: u8 = 0b0011_1110;
+    pub const OPCODE_CLK: u8 = 0b0011_1111;
+
+    pub const OPCODE_U32ADD: u8 = 0b0100_0000;
+    pub const OPCODE_U32SUB: u8 = 0b0100_0010;
+    pub const OPCODE_U32MUL: u8 = 0b0100_0100;
+    pub const OPCODE_U32DIV: u8 = 0b0100_0110;
+    pub const OPCODE_U32SPLIT: u8 = 0b0100_1000;
+    pub const OPCODE_U32ASSERT2: u8 = 0b0100_1010;
+    pub const OPCODE_U32ADD3: u8 = 0b0100_1100;
+    pub const OPCODE_U32MADD: u8 = 0b0100_1110;
+
+    pub const OPCODE_HPERM: u8 = 0b0101_0000;
+    pub const OPCODE_MPVERIFY: u8 = 0b0101_0001;
+    pub const OPCODE_PIPE: u8 = 0b0101_0010;
+    pub const OPCODE_MSTREAM: u8 = 0b0101_0011;
+    pub const OPCODE_SPLIT: u8 = 0b0101_0100;
+    pub const OPCODE_LOOP: u8 = 0b0101_0101;
+    pub const OPCODE_SPAN: u8 = 0b0101_0110;
+    pub const OPCODE_JOIN: u8 = 0b0101_0111;
+    pub const OPCODE_DYN: u8 = 0b0101_1000;
+    pub const OPCODE_RCOMBBASE: u8 = 0b0101_1001;
+
+    pub const OPCODE_MRUPDATE: u8 = 0b0110_0000;
+    pub const OPCODE_PUSH: u8 = 0b0110_0100;
+    pub const OPCODE_SYSCALL: u8 = 0b0110_1000;
+    pub const OPCODE_CALL: u8 = 0b0110_1100;
+    pub const OPCODE_END: u8 = 0b0111_0000;
+    pub const OPCODE_REPEAT: u8 = 0b0111_0100;
+    pub const OPCODE_RESPAN: u8 = 0b0111_1000;
+    pub const OPCODE_HALT: u8 = 0b0111_1100;
+}
+
 // OPERATIONS
 // ================================================================================================
 
-/// A set of native VM operations.
+/// A set of native VM operations which take exactly one cycle to execute.
 ///
-/// These operations take exactly one cycle to execute.
+/// Opcode patterns have the following meanings:
+/// - 00xxxxx operations do not shift the stack; constraint degree can be up to 2.
+/// - 010xxxx operations shift the stack the left; constraint degree can be up to 2.
+/// - 011xxxx operations shift the stack to the right; constraint degree can be up to 2.
+/// - 100xxx-: operations consume 4 range checks; constraint degree can be up to 3. These are used
+///   to encode most u32 operations.
+/// - 101xxx-: operations where constraint degree can be up to 3. These include control flow
+///   operations and some other operations requiring high degree constraints.
+/// - 11xxx--: operations where constraint degree can be up to 5. These include control flow
+///   operations and some other operations requiring very high degree constraints.
```

Comment:
I would probably move this explanation to line 14 above (i.e., for the constants module) as we see the actual opcode values there.

Reply:
done